stablehlo/dialect/Base.cpp

```diff
@@ -381,6 +381,52 @@ FailureOr<Type> inferMostSpecificType(std::optional<Location> location,
                                inferMostSpecificDimAndBound);
 }
 
+FailureOr<Type> mapOverTupleElements(
+    std::optional<Location> location, TypeRange inputTypes,
+    function_ref<FailureOr<Type>(std::optional<Location>, TypeRange types)>
+        fn) {
+  SmallVector<TupleType> tupleTypes;
+  for (auto inputType : inputTypes) {
+    if (auto tupleType = inputType.dyn_cast<TupleType>())
+      tupleTypes.push_back(tupleType);
+  }
+  if (!tupleTypes.empty()) {
+    if (tupleTypes.size() != inputTypes.size())
```

Comment:
Why does this restriction exist? If we are enforcing a contract, should it be "either must be a single tuple or all non-tuples" ?

Reply:
The idea behind this function is to generalize the existing logic of `inferMostSpecificType` and `inferLeastSpecificType`, so that these functions can be called on tuple types, with the original logic applied to these tuples elementwise and results packaged into a tuple type. For example:
```
inferMostSpecificType(tuple<T11, T12>, tuple<T21, T22>) =
  tuple<inferMostSpecificType(T11, T21), inferMostSpecificType(T12, T22)>
```